src/goto-instrument/contracts/contracts.cpp

```diff
@@ -686,106 +685,61 @@ void code_contractst::instrument_assign_statement(
+  else if(called_function == "free")
   {
     const exprt &lhs_dereference = dereference_exprt(
-      to_typecast_expr(instruction_iterator->call_arguments().front()).op());
-    add_containment_check(
-      program, assigns_clause, instruction_iterator, lhs_dereference);
-    assigns_clause.remove_from_local_write_set(lhs_dereference);
-    assigns_clause.remove_from_global_write_set(lhs_dereference);
+      to_typecast_expr(instruction_it->call_arguments().front()).op());
+    add_containment_check(body, assigns, instruction_it, lhs_dereference);
+    assigns.remove_from_local_write_set(lhs_dereference);
+    assigns.remove_from_global_write_set(lhs_dereference);
     return;
   }
 
   if(
-    instruction_iterator->call_lhs().is_not_nil() &&
-    instruction_iterator->call_lhs().id() == ID_symbol)
+    instruction_it->call_lhs().is_not_nil() &&
+    instruction_it->call_lhs().source_location().is_not_nil() &&
+    is_lvalue(instruction_it->call_lhs()))
   {
     add_containment_check(
-      program,
-      assigns_clause,
-      instruction_iterator,
-      instruction_iterator->call_lhs());
-  }
-
-  const symbolt &called_symbol = ns.lookup(called_name);
-  // Called symbol might be a function pointer.
-  const typet &called_symbol_type = (called_symbol.type.id() == ID_pointer)
-                                      ? called_symbol.type.subtype()
-                                      : called_symbol.type;
-  exprt called_assigns =
-    to_code_with_contract_type(called_symbol_type).assigns();
-  const code_typet &called_type = to_code_type(called_symbol_type);
-
-  if(called_assigns.is_not_nil())
-  {
-    replace_symbolt replace_formal_params;
-    const auto &arguments = instruction_iterator->call_arguments();
-    auto a_it = arguments.begin();
-    for(auto p_it = called_type.parameters().begin();
-        p_it != called_type.parameters().end() && a_it != arguments.end();
-        ++p_it, ++a_it)
-    {
-      if(!p_it->get_identifier().empty())
-      {
-        symbol_exprt p(p_it->get_identifier(), p_it->type());
-        replace_formal_params.insert(p, *a_it);
-      }
-    }
-    replace_formal_params(called_assigns);
-
-    // check subset relationship of assigns clause for called function
-    assigns_clauset called_assigns_clause(called_assigns, log, ns);
-    const auto subset_check =
-      assigns_clause.generate_subset_check(called_assigns_clause);
-    goto_programt alias_assertion;
-    alias_assertion.add(goto_programt::make_assertion(
-      subset_check, instruction_iterator->source_location));
-    alias_assertion.instructions.back().source_location.set_comment(
-      "Check that " + id2string(called_name) +
-      "'s assigns clause is a subset of " +
-      id2string(instruction_iterator->source_location.get_function()) +
-      "'s assigns clause");
-    insert_before_swap_and_advance(
-      program, instruction_iterator, alias_assertion);
+      body, assigns, instruction_it, instruction_it->call_lhs());
```

Comment:
We need a test to cover.

Reply:
Same as https://github.com/diffblue/cbmc/pull/6371#discussion_r720413355